### ISSUE TITLE
var(dim) and std(dim) are broken for FloatTensor

### ISSUE THREAD

**Issue by dlee** [pytorch/pytorch#2470]
It seems like `var(dim)` and `std(dim)` are broken on version 0.2.0.1. Please note that you need to pass in the `dim` argument to reproduce this bug. The parameterless `var()` and `std()` are working correctly.

For example:
```python
import torch
tensor = torch.FloatTensor([2281.5, 2281.25])
tensor.var(0) # => 0
tensor.var()  # => 0.03125
```

**Comment by CDLuminate** [comment]
Reproduced with `0.2.0_1`, and DoubleTensor is ok.

**Comment by fmassa** [comment]
I'd say this is due to precision errors during the computations, which are performed in single precision in the `var(0)` case, and in double precision in the `var()` case.
To inspect that, just perform the computations one by one on the terminal for example:
```python
# corresponds to the biased estimator
def var_all(t):
    s = t.sum() / t.numel()
    s2 = (t * t).sum() / t.numel()
    print('E[t]^2:', s * s)
    print('E[t^2]:', s2)
    return max((s2 - s * s), 0)

tensor = torch.DoubleTensor([2281.5, 2281.25])
print(var_all(tensor.float()))
print('-------------')
print(var_all(tensor))
```
Prints
```
E[t]^2: 5204671.890625
E[t^2]: 5204671.75
0
-------------
E[t]^2: 5204671.890625
E[t^2]: 5204671.90625
0.015625
```

But, looking at the code, it seems that the accumulation [is done in `double` precision](https://github.com/pytorch/pytorch/blob/master/torch/lib/TH/generic/THTensorMath.c#L2996-L2997), so I'm not sure why this is happening. I'm probably missing something obvious here

**Comment by pavanky** [comment]
@fmassa The final result of sum could still be rounded to the closest 32 bit float approximation.

**Comment by pavanky** [comment]
One alternative would be to use more stable versions of variance: 
https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance

**Comment by ngimel** [comment]
iirc, accreal type for float is float, so the computation is done in float. At any rate, the method used is extremely unstable, Welford method would give much better results https://docs.google.com/document/d/1V3VEm9zQ17jsbWT9dxK_iS5qRZZenU6tdApj-VsbGk0/edit

**Comment by fmassa** [comment]
@ngimel I thought that `accreal` in TH was `double` for `float`, and [it seems to be the case](https://github.com/pytorch/pytorch/blob/master/torch/lib/TH/THGenerateFloatType.h#L6) (even though in CUDA it's `float`). Anyway, the computation is indeed a bit unstable and might be better to improve it.

**Comment by ngimel** [comment]
Ah, ok, I mixed up with cuda. A bit unsettling that cuda and cpu results are expected to be different.

**Comment by ngimel** [comment]
varall uses two-pass algorithm to calculate variance (calculate mean, subtract, calculate variance), as opposed to var(dim), which uses an unstable one-pass: https://github.com/pytorch/pytorch/blob/master/torch/lib/TH/generic/THTensorMath.c#L3146-L3153, that would explain the difference between var() and var(dim). Welford (one-pass) algorithm should be better than even two-pass.

**Comment by ngimel** [comment]
Actually, std/var implementations are inconsistent both in TH and THC, for tensors and for variables.
For tensors:
1) in both TH and THC .var() is doing a two-pass variance calculation.
2) in both TH and THC .var(dim) is doing one-pass unstable variance calculation, however, in TH it accumulates in double, in THC it accumulates in float for float inputs. Even with double accumulation, it is really unstable, as original poster has shown. For half type in cuda, it will be truncated to half at each step, so it is basically unusable. In cuda, special kernels are used for that.
For variables all the std calculations are done using two-pass algorithm, which is better than one-pass naive, but still worse than one-pass Welford. It is implemented via tensor ops (calculate mean, subtract mean, do pointwise multiply, reduce). For cuda, it also results in bad performance, because 5-7 kernels and a coupe of synchronous d2h transfers are launched for each variance computation. The following script compares tensor/var performance on cuda, with 3x-4x slower perf for variables.
```.py
import torch
import time
from torch.autograd import Variable


def timeit(fn, nrep, args=None):
    torch.cuda.synchronize()
    start = time.time()
    for i in range(nrep):
       if args is not None:
           fn(args)
       else:
           fn()
    torch.cuda.synchronize()
    print((time.time()-start)*1000./float(nrep), "ms")

a=torch.randn(512,2048).cuda()
nrep = 100
timeit(a.var, nrep)
timeit(a.var, nrep, 0)
timeit(a.var, nrep, 1)
a=Variable(a)
timeit(a.var, nrep)
timeit(a.var, nrep, 0)
timeit(a.var, nrep, 1)
```
Also, almost everywhere mean and std are computed together, so mean computation is essentially repeated when std is computed. So the questions here are
1) does it make sense to have mean_std api to compute mean and variance/std together?
2) does it make sense to have special implementation for variables the way there are implementations now for tensors (though we cannot reuse current tensor implementations because they are broken)? It would also require special kernels for backward. It should help weight norm/layer norm layers.

**Comment by allenye0119** [comment]
I've also noticed some weird behavior of std(dim) on FloatTensor, maybe it's related to this thread. Running the following code:
```python
for i in range(16):
    x = 10 ** i
    t = torch.FloatTensor([x, x])
    std = t.std()
    std_0 = t.std(0)
    print('t.std(): {}'.format(std))
    print('t.std(0): {}'.format(std_0))
```
The output of std(0) will 'sometime' be nonzero.
